From 0122245f02ba7b22dd40f38c98aa2f08984707bd Mon Sep 17 00:00:00 2001
From: Henri Yandell <bayard@apache.org>
Date: Wed, 5 Nov 2008 15:47:43 +0000
Subject: [PATCH] Applying Bjorn Townsend's unit test and my fix for
 COLLECTIONS-304 - fixing SetUniqueList so the set method doesn't let the
 uniqueness get out of sync

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/collections/trunk@711591 13f79535-47bb-0310-9956-ffa450edef68
---
 .../collections/list/SetUniqueList.java       | 14 +++++-----
 .../collections/list/TestSetUniqueList.java   | 27 +++++++++++++++++++
 2 files changed, 35 insertions(+), 6 deletions(-)

diff --git a/src/java/org/apache/commons/collections/list/SetUniqueList.java b/src/java/org/apache/commons/collections/list/SetUniqueList.java
index 25d1504f09..8ad3553838 100644
--- a/src/java/org/apache/commons/collections/list/SetUniqueList.java
+++ b/src/java/org/apache/commons/collections/list/SetUniqueList.java
@@ -220,14 +220,16 @@ public boolean addAll(int index, Collection coll) {
     public Object set(int index, Object object) {
         int pos = indexOf(object);
         Object removed = super.set(index, object);
-        if (pos == -1 || pos == index) {
-            return removed;
+
+        if (pos != -1 && pos != index) {
+            // the object is already in the uniq list
+            // (and it hasn't been swapped with itself)
+            super.remove(pos);  // remove the duplicate by index
         }
-        
-        // the object is already in the uniq list
-        // (and it hasn't been swapped with itself)
-        super.remove(pos);  // remove the duplicate by index
+
+        set.add(object);      // add the new item to the unique set
         set.remove(removed);  // remove the item deleted by the set
+
         return removed;  // return the item deleted by the set
     }
 
diff --git a/src/test/org/apache/commons/collections/list/TestSetUniqueList.java b/src/test/org/apache/commons/collections/list/TestSetUniqueList.java
index ac68d72df3..cc82f479b3 100644
--- a/src/test/org/apache/commons/collections/list/TestSetUniqueList.java
+++ b/src/test/org/apache/commons/collections/list/TestSetUniqueList.java
@@ -443,4 +443,31 @@ public String getCompatibilityVersion() {
 //        writeExternalFormToDisk((java.io.Serializable) collection, "D:/dev/collections/data/test/SetUniqueList.fullCollection.version3.1.obj");
 //    }
 
+    public void testCollections304() {
+        List list = new LinkedList();
+        SetUniqueList decoratedList = SetUniqueList.decorate(list);
+        String s1 = "Apple";
+        String s2 = "Lemon";
+        String s3 = "Orange";
+        String s4 = "Strawberry";
+
+        decoratedList.add(s1);
+        decoratedList.add(s2);
+        decoratedList.add(s3);
+
+        assertEquals(3, decoratedList.size());
+
+        decoratedList.set(1, s4);
+
+        assertEquals(3, decoratedList.size());
+
+        decoratedList.add(1, s4);
+
+        assertEquals(3, decoratedList.size());
+
+        decoratedList.add(1, s2);
+
+        assertEquals(4, decoratedList.size());
+    }
+
 }
